dbms/src/Storages/Transaction/SSTReader.h

```diff
@@ -14,28 +14,110 @@
+    virtual BaseBuffView value() const;
+    virtual void next();
 
     DISALLOW_COPY_AND_MOVE(SSTReader);
     SSTReader(const TiFlashRaftProxyHelper * proxy_helper_, SSTView view);
-    ~SSTReader();
+    // Should only be used when SSTReader is a base class.
+    SSTReader();
+    virtual ~SSTReader();
 
 private:
     const TiFlashRaftProxyHelper * proxy_helper;
     SSTReaderPtr inner;
     ColumnFamilyType type;
+    bool inited;
 };
 
+/// MultiSSTReader helps when there are multiple sst files in a column family.
+/// It is derived from virtual class SSTReader, so it can be holded in a SSTReaderPtr.
+/// It also maintains instance of `R` which is normaly SSTReader(and MockSSTReader in tests),
+/// to read a single sst file.
+/// An `Initer` function is need to create a instance of `R` from a instance of `E`,
+/// which is usually path of the SST file.
+/// We introduce `R` and `E` to make this class testable.
+template <typename R, typename E>
+struct MultiSSTReader : SSTReader
+{
+    using Initer = std::function<std::unique_ptr<R>(const TiFlashRaftProxyHelper *, E)>;
+
+    DISALLOW_COPY_AND_MOVE(MultiSSTReader);
+
+    bool remained() const override
+    {
+        this->maybe_next_reader();
+        return mono->remained();
+    }
+    BaseBuffView key() const override
+    {
+        return mono->key();
+    }
+    BaseBuffView value() const override
+    {
+        return mono->value();
+    }
+    void next() override
+    {
+        this->maybe_next_reader();
+        mono->next();
+    }
+    void maybe_next_reader() const
+    {
+        if (!mono->remained())
+        {
+            current++;
+            if (current < args.size())
+            {
+                // We delete current mono, iif:
+                // 1. We can switch to next reader mono;
+                // 2. We must switch to next reader mono;
+                // We don't drop if mono is the last instance for safety,
+                // and it will be dropped as MultiSSTReader is dropped.
+                LOG_FMT_INFO(log, "Swithch to {}", buffToStrView(args[current].path));
+                mono = initer(proxy_helper, args[current]);
+            }
+        }
+    }
+
+    MultiSSTReader(const TiFlashRaftProxyHelper * proxy_helper_, ColumnFamilyType type_, Initer initer_, std::vector<E> args_)
+        : proxy_helper(proxy_helper_)
+        , type(type_)
+        , initer(initer_)
+        , args(args_)
+        , current(0)
+    {
+        assert(args.size() > 0);
+        log = &Poco::Logger::get("MultiSSTReader");
+        LOG_FMT_INFO(log, "Init with {}", buffToStrView(args[current].path));
+        mono = initer(proxy_helper, args[current]);
+    }
+
+    ~MultiSSTReader()
+    {
+        // The last sst reader will be dropped with inner.
+    }
+
+private:
+    Poco::Logger * log;
+    mutable std::unique_ptr<R> mono;
+    // Overwrite
+    const TiFlashRaftProxyHelper * proxy_helper;
+    // Overwrite
+    ColumnFamilyType type;
```

Comment:
make `SSTReader.proxy_helper` and `SSTReader.type` to be protected?

Reply:
Maybe it is better the base class exposed no its fields to either outside or its derived classes. Since `MultiSSTReader` inherits `SSTReader` only to implement its interfaces. So I think it is safer that we hide all fields of these classes by setting them private.

Reply:
I think defining members with duplicated names in the derived classes is a worse idea. If you want to make the fields not exposed its private method, then you can define some protect methods for read-only access to these fields.

Reply:
What I want is the derived class `MultiSSTReader` to have no access to all fields of its base class `SSTReader`, which is the reason why declaring them as private. Meanwhile, I find it hard to find another name for `type` and `proxy_helper` in derived class. So if `MultiSSTReader` already has no access to `SSTReader::tupe/proxy_helper`, we can define our own.

Reply:
OK, I know what the point is now. My mainly concern it make `MultiSSTReader` aware of nothing in `SSTReader` other than interfaces. This surely is a waste of memory, but only two integers in a limited used class. 
However, I think by totally deprivation of `SSTReader`, `MultiSSTReader` can no longer care about how `SSTReader` is implemented.